versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/Persist.java

```diff
@@ -0,0 +1,413 @@
+
+  /**
+   * Atomically deletes a reference, only available from <em>{@link #isTransactional()
+   * transactional}</em> implementations.
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   *
+   * @throws RefNotFoundException if a reference with the same name does not exist
+   * @throws RefConditionFailedException if the current pointer is different
+   */
+  default void deleteReference(@Nonnull Reference reference)
+      throws RefNotFoundException, RefConditionFailedException {
+    throw new UnsupportedOperationException(
+        "non-transactional implementations do not support Persist.deleteReference()");
+  }
+
+  /**
+   * Low-level, atomically updates the given reference's {@link Reference#pointer()} to the new
+   * value, if and only if the current persisted reference is not marked as {@link
+   * Reference#deleted()} and {@link Reference#pointer()} of the given and persisted values are
+   * equal.
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   *
+   * @return the updated {@link Reference}, if the reference exists, is not marked as {@link
+   *     Reference#deleted() deleted} and the {@link Reference#pointer()} update succeeded.
+   * @throws RefNotFoundException if a reference with the same name does not exist
+   * @throws RefConditionFailedException if the existing reference is marked as deleted or its
+   *     pointer is different
+   */
+  @Nonnull
+  Reference updateReferencePointer(@Nonnull Reference reference, @Nonnull ObjId newPointer)
+      throws RefNotFoundException, RefConditionFailedException;
+
+  /**
+   * Low-level, find a reference.
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   *
+   * @return the reference or {@code null}, if it does not exist
+   */
+  Reference findReference(@Nonnull String name);
+
+  /**
+   * Like {@link #findReference(String)}, but finds multiple references by name at once, leveraging
+   * bulk queries against databases.
+   *
+   * <p>Non-existing references are returned as {@code null} elements in the returned array.
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   */
+  @Nonnull
+  Reference[] findReferences(@Nonnull String[] names);
+
+  /**
+   * Lists references, starting at the given reference name, only available from <em>{@link
+   * #isTransactional() transactional}</em> implementations.
+   *
+   * <p>Note: this lists all references with reference names greater than or equal to {@code begin},
+   * which is different from {@link ReferenceLogic#queryReferences(ReferencesQuery)}, which performs
+   * a <em>prefix</em> check.
+   */
+  default CloseableIterator<Reference> listReferences(String begin) {
+    throw new UnsupportedOperationException(
```

Comment:
Would it be worth introducing a sub-interface for Tx persistence implementations?

Reply:
Not sure yet - two reasons:
* I did not want to have a "interface hierarchy nightmare" (as we have with the `*Config` interfaces in DA)
* Don't really know whether it's really a good idea to have a different model for Tx (up for discussion)

Reply:
> Don't really know whether it's really a good idea to have a different model for Tx (up for discussion)

If the distinction is mostly for a transactional list of references vs. a commit log-based list of references (not sure now - I've not reviewed all of the code yet). I'd think it might be preferable to have the same model for both backend types.

Reply:
I could imagine that _listing_ all references should be quicker w/ the non-tx model - i.e. having our own index at hand from a single row feels quicker / more efficient than listing many database rows. IDK

Reply:
I could imagine that _listing_ all references should be quicker w/ the non-tx model - i.e. having our own index at hand from a single row feels quicker / more efficient than listing many database rows. IDK